overlord/devicestate/handlers_install.go

```diff
@@ -618,3 +634,72 @@ func (m *DeviceManager) doRestartSystemToRunMode(t *state.Task, _ *tomb.Tomb) er
 
 	return nil
 }
+
+var seedOpen = seed.Open
+
+var maybeApplyPreseededData = func(ubuntuSeedDir, sysLabel, writableDir string) (preseeded bool, err error) {
+	preseedArtifact := filepath.Join(ubuntuSeedDir, "systems", sysLabel, "preseed.tgz")
+	if !osutil.FileExists(preseedArtifact) {
+		return false, nil
+	}
+
+	// TODO: verify artifact signature?
+
+	logger.Noticef("apply preseed data: %q, %q", writableDir, preseedArtifact)
+	cmd := exec.Command("tar", "--extract", "--preserve-permissions", "--preserve-order", "--gunzip", "--directory", writableDir, "-f", preseedArtifact)
+	if err := cmd.Run(); err != nil {
+		return false, err
+	}
+
+	logger.Noticef("copying snaps")
+
+	deviceSeed, err := seedOpen(dirs.SnapSeedDir, sysLabel)
+	if err != nil {
+		return false, err
+	}
+	tm := timings.New(nil)
+
+	// XXX: this is expensive, but we need snap revisions. Any way around this?
+	if err := deviceSeed.LoadAssertions(nil, nil); err != nil {
+		return false, err
+	}
+	if err := deviceSeed.LoadMeta(tm); err != nil {
+		return false, err
+	}
+
+	copyBlob := func(sn *seed.Snap) error {
+		sq := squashfs.New(sn.Path)
+		opts := &snap.InstallOptions{
+			MustNotCrossDevices: true,
+		}
+		pinfo := sn.PlaceInfo()
+		if pinfo.SnapRevision().Unset() {
+			pinfo = snap.MinimalPlaceInfo(pinfo.InstanceName(), snap.R(-1))
+		}
+		targetPath := filepath.Join(writableDir, pinfo.MountFile())
+		mountDir := filepath.Join(writableDir, pinfo.MountDir())
+		logger.Debugf("copying: %q to %q; mount dir=%q", sn.Path, targetPath, mountDir)
+		if _, err := sq.Install(targetPath, mountDir, opts); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	ess := deviceSeed.EssentialSnaps()
```

Comment:
nitpick^100: to be consistent with the naming below, this should be `esnaps` :-)